### PR TITLE
Major cleanup of SDK codegen

### PR DESCRIPTION
**Pull request by jkaster**
# OpenAPI conversion fixes

After using the open-source convert, fixups to merge things it missing from the swagger spec are performed and the converted spec file is updated.

- operation.responseBody.required is set correctly
- `x-looker-nullable` -> `nullable`
- convert parameter `collectionFormat` -> `style`
- all fixes are documented as part of the conversion process

# sdkModels.ts improvements and cleanup

Thanks to the improved OpenApi conversion, sdkModels has lots of improvements and clean up:

- no more need for the swagger spec!
- blazing fast specification search service (see sdkModels.spec.ts for details)
- type and method cross-referencing
  - `IType.methods` lists all methods that use the type
  - `IType.types` lists **all** the types used by the type
  - `IType.customTypes` lists all the API specification types
  - `IMethod.types` lists **all** the types used by the method
  - `IMethod.customTypes` lists all the API specification types used by the method

### CODE REVIEW

File: src/sdkModels.ts [hunk 1]
```diff
@@ -179,21 +246,20 @@ export interface IMethodResponse {
...
 
-  constructor(statusCode: string, mediaType: string, type: IType) {
-    this.statusCode = parseInt(statusCode, 10)
-    this.mediaType = mediaType
-    this.type = type
+  constructor(
+    public statusCode: number,
+    public mediaType: string,
+    public type: IType,
+    public description: string) {
```
Comment (joeldodge79) [inline src/sdkModels.ts:262]: slick!

File: src/convert.ts [hunk 2]
```diff
@@ -62,52 +64,158 @@ const lintCheck = async (fileName: string) => {
...
+  const swagger = JSON.parse(swaggerSpec)
+  const api = JSON.parse(openApiSpec)
+  const paths = swagger['paths']
+  const fixes: string[] = []
+  Object.entries(paths).forEach(([endpoint, op]) => {
+    Object.entries(op as any).forEach(([httpMethod, method]) => {
+      const operation = method as any
+      const params = operation.parameters
+      if (params) {
+        Object.entries(params).forEach(([, p], index) => {
```
[the remark this answers is not in the document]
Comment (jkaster) [reply src/convert.ts:128]: Yeah, index can be passed in with `Object.entries().forEach()`. For `for ...` construct is a fair point for readability. Object.entries() is more "compatible" between TS and JS I think, but IMO as well the `for ...` syntax is more readable. I may change it although the functionality will be the same.

File: src/convert.ts [hunk 3]
```diff
@@ -62,52 +64,158 @@ const lintCheck = async (fileName: string) => {
...
+          const param = p as any
+          if (param.name === 'body' && param.in === 'body') {
+            // Set `required` in requestBody
+            if ('required' in param) {
+              //  explicitly setting required value
+              const required = param.required
+              const fix = `${endpoint} ${operation.operationId} setting requestBody.required to ${required}`
+              const body = api.paths[endpoint][httpMethod].requestBody
+              if (!body) {
+                warn(`Failed, not found: ${fix}`)
```
[the remark this answers is not in the document]
Comment (jkaster) [reply src/convert.ts:138]: Good suggestion.

File: src/convert.ts [hunk 4]
```diff
@@ -62,52 +64,158 @@ const lintCheck = async (fileName: string) => {
...
+ * @param {string} collectionFormat
+ * @returns {OpenApiStyle}
+ */
+export const openApiStyle = (collectionFormat: string): OpenApiStyle => {
+  if (!collectionFormat) return undefined
+  const styles: {[key:string]: OpenApiStyle} = {
+    'csv': 'simple',
+    'ssv': 'spaceDelimited',
+    'pipes': 'pipeDelimited',
+  }
```
[the remark this answers is not in the document]
Comment (jkaster) [reply src/convert.ts:93]: Nice tweak but I'm not sure it adds any value here and would require double entry if we end up finding another special delimited format to support. The fall through code handles the unrecognized condition, as is intended.

directly below that is:
```typescript
  if (collectionFormat in styles) {
    return styles[collectionFormat]
  }
  return undefined
```

File: src/sdkModels.ts (hunk 1 above)
Comment (jkaster) [reply src/sdkModels.ts:262]: Are you talking about Typescript's syntactic sugar to declare a constructor parameter as a public property?

File: src/reformatter.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
+/*
```
[the remark this answers is not in the document]
Comment (jkaster) [reply src/reformatter.ts:1]: What's handy about this is we can very easily create a general purpose source code reformatter script from this that handles all our supported languages ... as we add reformatters for those languages (Kotlin and Swift) that we don't currently support, too. But for Typescript and Python we could handle an automatic recursive path reformatter, and only those files that are recognized would be part of the reformat request.

File: src/convert.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (joeldodge79) [reply src/convert.ts:93]: if `'csv'|'ssv'|'pipes'` is too limiting I think the following is equivalent to what you have:

File: src/sdkModels.ts (hunk 1 above)
Comment (joeldodge79) [reply src/sdkModels.ts:327]: that and the auto-assignment to those public members at instantiation time. nice to remove boilerplate

File: src/convert.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkaster) [reply src/convert.ts:93]: @joeldodge79 did you mean to type something after the `:` above?

[the remark this answers is not in the document]
Comment (joeldodge79) [reply src/convert.ts:93]: oops, ignore that comment, I thought I'd deleted it. I asked the internet about using this sort of "inline interface declaration" (if that's even an accurate label?) vs a `Record` and I found one SO article saying it's idiomatically preferred over the equivalent `Record<string, OpenApiStyle>` I was going to suggest https://stackoverflow.com/questions/51936369/what-is-the-record-type-in-typescript

[the remark this answers is not in the document]
Comment (jkaster) [reply src/convert.ts:93]: Good discussion to review. Thanks for the follow-up!

**Pull request merged**